Add `EntityMut::remove_by_id` and `EntityCommands::remove_by_id` methods

mateuseap opened the pull request:
# Objective

- Add missing methods to `EntityCommands` and `EntityMut`.
- Fix #9261 

## Solution

- I have checked out some existing code and also based myself on the useful informations provided in this [comment](https://github.com/bevyengine/bevy/issues/9261#issuecomment-1697299482) by @Selene-Amanita to implement these three new methods. These methods have been developed to address specific requirements that were previously lacking.

---

## Changelog

- Added `EntityMut::remove_by_id` method and its tests.
- Added `EntityCommands::remove_by_id` method and its tests.
- Refactor `EntityCommands::insert` and `Insert` to use `EntityCommand`

Selene-Amanita commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -890,6 +890,70 @@ impl<'w> EntityWorldMut<'w> {
+    /// # Safety
+    ///
+    /// - [`ComponentId`] must be from the same world as [`EntityWorldMut`]
+    /// - [`OwningPtr`] must be a valid reference to the type represented by [`ComponentId`]
```
```suggestion
```
There is no OwningPtr

Selene-Amanita commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -890,6 +890,70 @@ impl<'w> EntityWorldMut<'w> {
+        let removed_components = &mut self.world.removed_components;
+
+        let bundles = &mut self.world.bundles;
+        let bundle_info = bundles.init_info::<EmptyBundle>(components, storages);
```
I don't think it works to just put `EmptyBundle` here? I think you need to do something similar to `insert_by_id` and get bundle_info via `self.world.bundles.init_component_info(self.world.components, component_id)`.

The way you're doing it now I suspect you would end up with the same archetype and return line 925.

Selene-Amanita commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -890,6 +890,70 @@ impl<'w> EntityWorldMut<'w> {
+        }
+
+        #[allow(clippy::undocumented_unsafe_blocks)] // TODO: document why this is safe
+        unsafe {
```
This is already an unsafe context you can probably remove this (but maybe keep the comment about safety).

Selene-Amanita commented on crates/bevy_ecs/src/system/commands/mod.rs:
```diff
@@ -722,6 +723,16 @@ impl<'w, 's, 'a> EntityCommands<'w, 's, 'a> {
 
+    /// Adds a [`Component`] to the entity.
+    pub fn insert_by_id(&mut self, id: ComponentId, component: impl Component) -> &mut Self {
+        self.commands.add(InsertById {
```
```suggestion
        self.add(InsertById {
```
I think you can do that if `InsertById` implements `EntityCommand` instead?
I'm confused why `Insert` doesn't do that tbh, is it legacy code?
Maybe you can change that for `Insert` too, and for the other commands that have an `Entity` field?
If there are a lot of them don't bother.

mateuseap replied:
I've removed this `.commands` here and also in other commands that have an Entity field! (I'm not sure if I've removed everything)

Selene-Amanita commented on crates/bevy_ecs/src/system/commands/mod.rs:
```diff
@@ -959,6 +979,27 @@ where
+impl Command for InsertById {
+    fn apply(self, world: &mut World) {
+        if let Some(mut entity) = world.get_entity_mut(self.entity) {
+            entity.insert_by_id(self.id, self.component);
```
Shouldn't this be inside an `unsafe` block?
I think you need to verify the safety constraints are valid before calling it too.
For `remove_by_id` you can just check that `self.id` is indeed a valid component id, but for `InsertById` you need to verify that `self.component` has the right underlying type and I'm not sure how to do that.

mateuseap replied:
I've added a `unsafe` block and implemented the component type validation, but I'm not sure if it's right. Also, the stuff related to `EntityCommands::insert_by_id` method is not right yet, there still a couple of errors

Selene-Amanita reviewed (commented):
`remove_by_id` needs a test like `entity_mut_insert_by_id`.

mateuseap commented:
@Selene-Amanita, I've made the changes you recommended! I'd really appreciate it if you could take another look at my code and provide further guidance on how to make it more effective. I'm still encountering issues related to the `EntityCommands::insert_by_id` method. Your assistance would be greatly appreciated.

Selene-Amanita reviewed (commented):
I'm not sure how to deal with `StorageType` / lifetime for `InsertById` honestly, sorry :(

You can mark other "requested/recommended changes" conversations as "resolved".

SkiFire13 commented:
I'm not sure why `insert_by_id` is needed. If you already have an `impl Component` can't you just call `insert`?

Selene-Amanita commented:
> I'm not sure why `insert_by_id` is needed. If you already have an `impl Component` can't you just call `insert`?

You might not know the type of the component at compile time, think of a scene loading from file situation, or loading a GLTF with extra information on which components to add.

`insert_by_id` already existed on `EntityMut`, this PR adds it on `EntityCommands`, and adds `remove_by_id` on both.

SkiFire13 commented:
> You might not know the type of the component at compile time

But the given function clearly takes an `impl Component`, which requires you to know its type!

> `insert_by_id` already existed on `EntityMut`, this PR adds it on `EntityCommands`

`EntityMut::insert_by_id` is `unsafe` and takes a `OwningPtr<'_>` rather than an `impl Component` though.

Selene-Amanita commented:
@SkiFire13 Fair. Is there a way to have a safe `insert_by_id` method on `EntityCommands` then? Can you just use an `OwningPtr` too?

SkiFire13 commented:
I don't think there's a way to have a safe `insert_by_id`. For it to be safe the command would have to be able to check that the `ComponentId` passed is the correct one (like you're doing right now), but this inevitably requires it to know the correct `ComponentId`, so there's no point to passing one yourself.

But even keeping it `unsafe`, I don't think it can take an `OwningPtr<'_>` because it doesn't own the backing storage, only the value contained in it. This is why it has a lifetime, because it borrows the storage from somewhere (potentially a local value of the system). This means the borrow will most likely not be valid when the command is executed, so it can't capture the `OwningPtr<'_>`. You would have to find some other storage for it, but given that you can't know the size/alignment of a `OwningPtr<'_>` alone, the user would have to give it yourself or they would have to move the value from a `OwningPtr<'_>` to some other form of owned storage.

IMO this is too complex and far too uncommon for a command, and if someone really needs it they can just make their own.

Selene-Amanita commented:
@SkiFire13 thank you for the insight!

@mateuseap I think you can remove the `insert_by_id` on `EntityCommands` actually.

And because:
- it would be a bit weird to have `insert_by_id` and not `remove_by_id`,
- needing `remove_by_id` is a pretty rare use case,
- people can make their own `EntityCommand` if needed...

... I think maybe removing `remove_by_id` on `EntityCommands` and keeping it only on `EntityMut` would make more sense. I don't have a strong opinion on that though.

JoJoJet commented on crates/bevy_ecs/src/system/commands/mod.rs:
````diff
@@ -715,13 +716,23 @@ impl<'w, 's, 'a> EntityCommands<'w, 's, 'a> {
     /// # bevy_ecs::system::assert_is_system(add_combat_stats_system);
     /// ```
     pub fn insert(&mut self, bundle: impl Bundle) -> &mut Self {
-        self.commands.add(Insert {
+        self.add(Insert {
             entity: self.entity,
             bundle,
         });
````
If we're changing `Insert` to implement `EntityCommand` instead of `Command`, then we don't need to store the entity as a field anymore. It gets passed as an argument when `EntityCommand::apply` gets called.

SkiFire13 commented on crates/bevy_ecs/src/system/commands/mod.rs:
```diff
@@ -990,6 +990,25 @@ impl<T> Remove<T> {
+impl EntityCommand for RemoveById {
+    fn apply(self, id: Entity, world: &mut World) {
+        if let Some(mut entity_mut) = world.get_entity_mut(id) {
+            // SAFETY: `RemoveById` is only created by `EntityCommands::remove_by_id`, which
+            // ensures that the component exists on the entity.
```
`EntityCommands::remove_by_id` doesn't seem to check anything though

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -890,6 +890,75 @@ impl<'w> EntityWorldMut<'w> {
+    ///
+    /// You should prefer to use the typed API [`EntityWorldMut::remove`] where possible.
+    ///
+    /// # Safety
+    ///
+    /// - [`ComponentId`] must be from the same world as [`EntityWorldMut`]
+    pub unsafe fn remove_by_id(&mut self, component_id: ComponentId) -> &mut Self {
+        let archetypes = &mut self.world.archetypes;
+        let storages = &mut self.world.storages;
+        let components = &mut self.world.components;
+        let entities = &mut self.world.entities;
+        let removed_components = &mut self.world.removed_components;
+
+        let (bundle_info, _storage_type) = self
+            .world
+            .bundles
+            .init_component_info(components, component_id);
+
+        let old_location = self.location;
+        // SAFETY: `archetype_id` exists because it is referenced in the old `EntityLocation` which is valid,
+        // components exist in `bundle_info` because `Bundles::init_info` initializes a `BundleInfo` containing all components of the bundle type `T`
+        let new_archetype_id = remove_bundle_from_archetype(
```
Can we make this safe? `Bundles::init_component_info` (and not `Bundles::init_info`, like the safety comment claims) panics if no component with id `component_id` exists, so `remove_bundle_from_archetype` should always be sound to call.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -890,6 +890,75 @@ impl<'w> EntityWorldMut<'w> {
+        )
+        .expect("intersections should always return a result");
+
+        if new_archetype_id == old_location.archetype_id {
+            return self;
+        }
+
+        let old_archetype = &mut archetypes[old_location.archetype_id];
+        let entity = self.entity;
+        if old_archetype.contains(component_id) {
```
If the archetype changed then this `contains` should always be `true`

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -890,6 +890,75 @@ impl<'w> EntityWorldMut<'w> {
+
+            // Make sure to drop components stored in sparse sets.
+            // Dense components are dropped later in `move_to_and_drop_missing_unchecked`.
+            if let Some(StorageType::SparseSet) = old_archetype.get_storage_type(component_id) {
```
You could check `_storage_type` here instead.

SkiFire13 commented on crates/bevy_ecs/src/system/commands/mod.rs:
```diff
@@ -940,21 +944,17 @@ impl Command for Despawn {
-        if let Some(mut entity) = world.get_entity_mut(self.entity) {
-            entity.insert(self.bundle);
-        } else {
-            panic!("error[B0003]: Could not insert a bundle (of type `{}`) for entity {:?} because it doesn't exist in this World.", std::any::type_name::<T>(), self.entity);
+    fn apply(self, id: Entity, world: &mut World) {
+        if let Some(mut entity_mut) = world.get_entity_mut(id) {
+            entity_mut.insert(self.bundle);
         }
```
Note that while `EntityCommands` checks that the entity exists when it is created, this doesn't mean the entity will still exist when the command is applied. This means that this is changing the behaviour of `insert` when an entity is despawned inbetween submitting the command and applying it.

hymm replied:
why is the panic removed?

mateuseap commented:
@Selene-Amanita @JoJoJet @SkiFire13 Thank you all for your valuable feedback!

Here's a summary of the updates made:
1. Removed `EntityCommands::insert_by_id`.
2. Modified `EntityCommands::remove_by_id` and `EntityMut::remove_by_id` methods.
3. Updated tests for `EntityCommands::remove_by_id`.
4. Formatted the code and refined some comments.

Additionally, I addressed @JoJoJet's suggestion for the `Insert` implementation:
- `Insert` now implements `EntityCommand`, eliminating the need to store the entity as a field. It's passed as an argument when `EntityCommand::apply` is called.

I appreciate the input from @SkiFire13, and I'm eager to hear your thoughts, @Selene-Amanita and @JoJoJet, before proceeding with any further changes. Please feel free to share any additional suggestions or feedback!

JoJoJet commented:
Overall I like the state of this PR. Once you address the comments by @SkiFire13 I think this should be ready to merge.

Also, you should add a small migration guide mentioning the change to `Insert`.

mrchantey commented:
@mateuseap are you on track to get these changes in for `0.14`? Otherwise I'd be happy to finish it up in another PR.

mrchantey commented:
With #12842 merged this can be closed 🎉
